design-proposals/klip-35-dynamic-processing-log-levels.md

```diff
@@ -0,0 +1,210 @@
+                         "properties" : {
+                             “level”: {“type”: “string”, “description”: “The level of the logger”}
+                         }
+                     }
+                 }
+            }
+        }
+    }
+}
+```
+
+### ksqlDB Statements
+
+As a convenience, in addition to the API we’ll include ksqlDB statements for getting/setting the level. This makes it easy to get/set the level from ksqlDB user interfaces (such as the Confluent Cloud or C3 UI) while those products implement their own interface to the processing log (if they choose to do so).
+
+- `SET_PLL root <level>`: sets the processing log level for the root logger
+- `UNSET_PLL root`: unsets the processing log level for the root logger
+- `GET_PLL`: lists all processing log levels.
+- `SET_PLL <query context> <level>`: sets the processing log level for the given logger
+- `UNSET_PLL <query context> <level>`: unsets the processing log level for the given logger
+- `GET_PLL <query context>`: lists all processing log levels with the given logger prefix
+
+## Design
+
+### Storing Log Levels
+
+The log levels will be stored on a Kafka topic called the processing log config topic (PLCT). The key of the topic’s records will be the processing logger name. We will use the name “root” for the root logger. The value of the topic will be a JSON object representing the logger’s config. For now, it will contain a single field called “level”:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+```
+
+The Kafka topic will be compacted.
+
+The name of the topic will be controlled by a configuration named “ksql.logging.processing.config.topic.name”.
+
+### Setting a Log Level
+
+To set the log level, the ksql node receiving the request to set the level will produce a record with the logger’s key and value containing the specified level to the PLCT. Each ksql node will run a thread that consumes the PLCT. For each message consumed, the thread will use the log4j API to set the log level. To set the root level, the thread will call:
+
+```
+LogManager.getRootLogger().setLevel(<level>);
+```
+
+To set a child logger level, the thread will call:
+
+```
+LogManager.getLogger().setLevel(<level>);
+```
+
+Unsetting a Log Level
+
+To unset the log level, the thread consuming the PLCT will set the log level for the specified logger to `null`.
+
+### Getting/Listing Log Levels
+
+To get/list log levels, the ksql node receiving the request will list the local loggers in log4j and use them to populate the response.
+
+## Test Plan
+In  addition to the usual unit testing, we’ll include an  integration test to validate that the processing log is populated correctly when the level is set.
+
+## LOEs
+- ksqlDB syntax: 3 days
+- API endpoints: 4 days
+- implementation of the backend: 5 days
+- adding node-level debug logging of the data seen at each node: 1 day
```

Comment:
What does "data" here refer to? The log levels being read from the config topic?

Reply:
no I mean logging the data seen at each node of query processing (e.g. reading a source, projecting, filtering, aggregation, etc) at a very verbose level (e.g. TRACE)

Reply:
Oh interesting. Is this a highly requested feature? (I was confused since the KLIP says adding more verbose log messages was out of scope for this KLIP.)

Reply:
Ah yeah I should clear that up. I have a larger set of useful messages in mind that would help debug some of the issues I saw in community slack. The idea here is to put some super basic logging in so that the feature is actually doing something.